torch/distributed/checkpoint/state_dict_loader.py

```diff
@@ -107,17 +107,24 @@ def load(
         rank has an individual GPU, via ``torch.cuda.set_device()``.
     """
 
+    keys = _all_gather_keys(state_dict)
```

Comment:
are we using object collectives here? perf is not great for those.

also we are introducing a strict requirement that users call state_dict on all ranks of the world at around the same time.  I suppose that is an OK assumption. 

is it invalid to save just part of a model?

Reply:
@fegin @wz337 correct me if I'm wrong, but I believe it's expected all ranks of the world are calling `load` around the same time since we have to `all_gather` the load plan at the same time https://github.com/pytorch/pytorch/blob/ec124b90b88061aca3a74277b4c8df47414221a2/torch/distributed/checkpoint/state_dict_loader.py#L110

Reply:
I think users would only need to call `state_dict` on all ranks if they are using collectives in the `state_dict` functions. 

Performance issues are def a valid point! Do we have any way to quantify how bad the drop is from this one `all_gather`? I had envisioned it not being a large number of keys since this is the outermost state dict and nothing is being flattened.

Reply:
This only gathers over the first level state_dict keys which are like "model", "optimizer", "trainer", "dataloader". And this is specific done only for loading not saving. So I guess the performance is not a problem and the assumption of all the ranks should participate should be fine as well. We have the same assumption for save as well.